staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -745,7 +815,7 @@ func (p *sharedProcessor) addListener(listener *processorListener) ResourceEvent
 	p.listeners[listener] = true
 
 	if p.listenersStarted {
-		p.wg.Start(listener.run)
+		p.wg.StartWithContext(ctx, listener.run)
 		p.wg.Start(listener.pop)
```

Comment:
> why StartWithContext for listener.run but not listener.pop (same question applies elsewhere)?

I need to double-check that.

> I'm finding it really confusing to understand when the threaded context should be used for cancellation in various places and when it's just being used for extracting a contextual logger. I expect lots of confusion and mistakes :-/

The alternative is to pass a `logger klog.Logger`. That has been done in the past elsewhere, only to find out later that the callchain needed a real context after all. I prefer passing a context in all exported APIs because that enables users of client-go to use slog by putting such a logger into the context (logr/klog and slog are interoperable).

That then leads to the situation that code which only has a logger must first construct a context when calling such public APIs.

Also, callers which have a context are forced to extract the logger when invoking a function which uses that in its prototype, even if that function then only uses the logger in error paths that are not reached in most cases.

> The only way I can think to avoid this is to document at every level what the context is used for, especially when it's explicitly not used for cancellation.

I agree that the "not used for cancellation" part is problematic. Perhaps we can avoid that problem by ensuring that code which takes a context also honors the cancellation, in addition to whatever other signals for termination it has?

I'll check that for this code here.

Reply:
I'm not sure why I didn't change `pop` the same way as `run`. Perhaps I made those changes at different times.

Now both `pop` and `run` take a context as parameter, and both return when that context is canceled: https://github.com/kubernetes/kubernetes/compare/566db2763057755e7ca3698e4ae443a903cd8969..f2dcad6300494ce13ee9e9a0d60b090c19aeaf52

Reply:
> same question applies elsewhere

Do I have other such inconsistencies?

Reply:
> The alternative is to pass a `logger klog.Logger`. That has been done in the past elsewhere, only to find out later that the callchain needed a real context after all. I prefer passing a context in all exported APIs because that enables users of client-go to use slog by putting such a logger into the context (logr/klog and slog are interoperable).
> 
> That then leads to the situation that code which only has a logger must first construct a context when calling such public APIs.

I think that making functions take contexts which *today* do not use them for cancellation means that we cannot later use them for cancellation without risking hanging invocations which passed in TODO/Background contexts because they were forced to by the signature.

> Perhaps we can avoid that problem by ensuring that code which takes a context also honors the cancellation, in addition to whatever other signals for termination it has?
> 
> I'll check that for this code here.

I'm torn between making this PR include even more non-mechanical changes which have to be reasoned through, and this PR leaving things in a sort of misleading state where contexts are plumbed through and it's unclear what they are intended to be used for.

Reply:
> I'm not sure why I didn't change `pop` the same way as `run`. Perhaps I made those changes at different times.
> 
> Now both `pop` and `run` take a context as parameter, and both return when that context is canceled: https://github.com/kubernetes/kubernetes/compare/566db2763057755e7ca3698e4ae443a903cd8969..f2dcad6300494ce13ee9e9a0d60b090c19aeaf52

What does it mean if `run` queues an event and `pop` never drains it because it canceled on the context? Is that a new hang / leak that this makes possible?

I'm really not trying to delay merge of this, I'm just trying to make sure we leave this in a state we can understand _and_ are sure doesn't regress current behavior.

Reply:
> I think that making functions take contexts which today do not use them for cancellation means that we cannot later use them for cancellation without risking hanging invocations which passed in TODO/Background contexts because they were forced to by the signature.

We cannot make the context the *only* method of canceling them, but we can *add* cancellation through that context. Suppose we have `foo(ctx context.Context)` today which does some blocking operation and returns when some event occurs, ignoring cancellation of that context. Some caller uses `foo(context.Background())`. That call returns as soon as it is meant to return (= event occurred).

Later we allow `foo` to also return when its context is canceled. Then `foo(context.Background())` remains valid and does the same as before.

A problem could be that if `foo` is passed a context which gets cancelled. That worked before (foo ignored cancellation) but won't work anymore once it does.

> I'm torn between making this PR include even more non-mechanical changes which have to be reasoned through

I suspect this PR can't be about purely non-mechanical changes :cry: 

> and this PR leaving things in a sort of misleading state where contexts are plumbed through and it's unclear what they are intended to be used for

My preference would be to ensure that if a context is passed, it is used for both contextual logging and cancellation. That is true now for `run` and `pop`.

I'm not sure about the rest of the changes. Let's decide about the goal and then I can check.

Reply:
> > I think that making functions take contexts which today do not use them for cancellation means that we cannot later use them for cancellation without risking hanging invocations which passed in TODO/Background contexts because they were forced to by the signature.
> 
> We cannot make the context the _only_ method of canceling them, but we can _add_ cancellation through that context.

I'm pretty sure that's not generally safe to change because of what you point out about passing a context that gets canceled. That seems like a real risk.

1. Constructor takes a context, stores it, does not use it for cancellation
2. user calls it like this:
   ```
   ctx, cancel := context.WithCancel(parentCtx)
   defer cancel()
   controller := foo.NewController(ctx)
   return controller
   ```
3. We switch something in the guts of the controller to cancel on the passed-in context

Now the user is constructing a controller that dies on return when it didn't before.

Reply:
My rationale was that if we don't say explicitly "this context is only for contextual logging, not for cancellation", then users should not rely on cancellation being ignored. But of course, documentation is one thing, actual behavior that user start to rely on another...

If we say it, then we have to stand by what we promise and indeed cannot change it later. This may be useful in APIs which have very special termination semantics.

My proposal still stands: let's avoid the ambiguity by ensuring that cancellation works from day one of adding an API with a context parameters, with only few carefully justified exceptions.

Reply:
> What does it mean if run queues an event and pop never drains it because it canceled on the context? Is that a new hang / leak that this makes possible?

Both `pop` and `run` get the same parent context from `AddEventHandlerWithContext`. If that context gets canceled, all processing shuts down and goroutines quit. The caller shouldn't expect any specific semantic, like "shut down cleanly" - whatever that would mean here.

Reply:
I'm not quite satisfied with this. I suspect that the processor might get stuck if a listener just stops without being removed. This doesn't matter if the processor and listener use the same root context and that gets canceled, but this is not guaranteed.

I'm undecided whether I should fix this (should be possible) or go back to "context cancellation is ignored" in `AddEventHandlerWithContext` (simpler). If we pick "is ignored", it becomes part of the API contract and cannot be changed, as discussed above. Any preference?

cc @soltysh